integration/container_volume_test.go

```diff
@@ -0,0 +1,151 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package integration
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+)
+
+func createRegularFile(basePath, content string) (string, error) {
+	newFolder := filepath.Join(basePath, "regular")
+	err := os.Mkdir(newFolder, 0755)
+	if err != nil {
+		return "", err
+	}
+
+	newFile := filepath.Join(newFolder, "foo.txt")
+	err = ioutil.WriteFile(newFile, []byte(content), 0644)
+	return filepath.Join("regular", "foo.txt"), err
+}
+
+func fileInSymlinkedFolder(basePath, targetFile string) (string, error) {
+	symlinkFolder := filepath.Join(basePath, "symlink_folder")
+	err := os.Symlink(filepath.Dir(targetFile), symlinkFolder)
+
+	return filepath.Join(symlinkFolder, filepath.Base(targetFile)), err
+}
+
+func symlinkedFile(basePath, targetFile string) (string, error) {
+	symlinkFile := filepath.Join(basePath, "symlink_file")
+	err := os.Symlink(targetFile, symlinkFile)
+
+	return symlinkFile, err
+}
+
+func symlinkedFileInSymlinkedFolder(basePath, targetFile string) (string, error) {
+	symlinkFolderFile, err := fileInSymlinkedFolder(basePath, targetFile)
+	if err != nil {
+		return "", err
+	}
+
+	return symlinkedFile(basePath, symlinkFolderFile)
+}
+
+func TestContainerSymlinkVolumes(t *testing.T) {
+	for name, testCase := range map[string]struct {
+		createFileFn func(basePath, targetFile string) (string, error)
+	}{
+		// Create difference file / symlink scenarios:
+		// - symlink_file -> regular_folder/regular_file
+		// - symlink_folder/regular_file (symlink_folder -> regular_folder)
+		// - symlink_file -> symlink_folder/regular_file (symlink_folder -> regular_folder)
+		"file in symlinked folder": {
+			createFileFn: fileInSymlinkedFolder,
+		},
+		"symlinked file": {
+			createFileFn: symlinkedFile,
+		},
+		"symlinkedFileInSymlinkedFolder": {
+			createFileFn: symlinkedFileInSymlinkedFolder,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
```

Comment:
I think may need to capture `testCase` here in a local variable

```suggestion
		testCase := testCase // capture range variable
		t.Run(name, func(t *testing.T) {
```

See https://blog.golang.org/subtests and https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables

Reply:
Sure. Although, I got this implementation from ``integration/pod_hostname_test.go`` and ``integration/container_stop_test.go``, and there might be a few other scenarios like this in ``integration/client``.